src/main/java/graphql/util/TraverserState.java

```diff
@@ -56,21 +65,27 @@ private QueueTraverserState(Object sharedContextData) {
         @Override
         public void pushAll(TraverserContext<U> traverserContext, Function<? super U, Map<String, ? extends List<U>>> getChildren) {
             Map<String, ? extends List<U>> childrenMap = getChildren.apply(traverserContext.thisNode());
+            Map<String, List<TraverserContext<U>>> childrenContextMap = new LinkedHashMap<>();
             childrenMap.keySet().forEach(key -> {
                 List<U> children = childrenMap.get(key);
                 for (int i = 0; i < children.size(); i++) {
                     U child = assertNotNull(children.get(i), "null child for key " + key);
                     NodePosition nodePosition = new NodePosition(key, i);
-                    super.state.add(super.newContext(child, traverserContext, nodePosition));
+                    DefaultTraverserContext<U> context = super.newContext(child, traverserContext, nodePosition);
+                    childrenContextMap.computeIfAbsent(key, notUsed -> new ArrayList<>());
+                    childrenContextMap.get(key).add(context);
+                    super.state.add(context);
                 }
             });
-            super.state.add(Marker.END_LIST);
+            EndList<U> endList = new EndList<>();
+            endList.childrenContextMap = childrenContextMap;
+            super.state.add(endList);
             super.state.add(traverserContext);
         }
     }
 
-    public enum Marker {
-        END_LIST
+    public static class EndList<U> {
+        public Map<String, List<TraverserContext<U>>> childrenContextMap;
     }
```

Comment:
whats up with this class??   A static class with a field with public access???  Not great API